src/mscorlib/src/System/String.cs

```diff
@@ -3542,18 +3533,42 @@ public static string Concat(params object[] args)
             }
 
             int totalLength = str0.Length + str1.Length + str2.Length + str3.Length;
+            string result = FastAllocateString(totalLength);
 
-            String result = FastAllocateString(totalLength);
-            FillStringChecked(result, 0, str0);
-            FillStringChecked(result, str0.Length, str1);
-            FillStringChecked(result, str0.Length + str1.Length, str2);
-            FillStringChecked(result, str0.Length + str1.Length + str2.Length, str3);
+            fixed (char* pResult = &result.m_firstChar)
+            {
+                str0.CopyToPointer(pResult);
+                str1.CopyToPointer(pResult + str0.Length);
+                str2.CopyToPointer(pResult + str0.Length + str1.Length);
+                str3.CopyToPointer(pResult + str0.Length + str1.Length + str2.Length);
+            }
 
             return result;
         }
 
+        // This is separated out into a different method since
+        // it seems to improve performance/code size for certain
+        // Concat() overloads as opposed to when it is written inline.
+        // Please do not use this method for other purposes unless
+        // you can be sure that it is benefiting your use case.
+        // In Concat's use case, this was beneficial since the result
+        // of fixed on the destination could be cached, while there
+        // were multiple sources that had to be fixed on.
+        // Please make thorough (repeated) microbenchmarks of how
+        // this will affect the Concat() overloads before making changes.
+        private unsafe void CopyToPointer(char* destination)
```

Comment:
@jkotas Are you saying that writing it inline does not make sense, or that putting it into a new method improves perf does not make sense? Not sure either way, so please enlighten me...